src/Pos/Wallet/Web/Server/Methods.hs

```diff
@@ -330,69 +367,128 @@ getUserProfile = getProfile >>= maybe noProfile pure
+            remAddr  <- decodeCAddressOrFail $ caaAddress remCAddr
+            let remTx = TxOutAux (TxOut remAddr $ balance `unsafeSubCoin` c) []
+            return $ Just (remTx, remCAddr)
     na <- getKnownPeers
     withSafeSigner sk (return passphrase) $ \ss -> do
-        etx <- submitTx sendActions ss na (one $ TxOutAux (TxOut dstAddr c) [])
+        etx <- submitTx sendActions ss na (mainTx :| maybe mempty (one . fst) mRems)
         case etx of
             Left err -> throwM . Internal $ sformat ("Cannot send transaction: "%stext) err
             Right (tx, _, _) -> do
                 logInfo $
-                    sformat ("Successfully sent "%coinF%" from "%ords%" address to "%addressF)
-                    c idx dstAddr
+                    sformat ("Successfully sent "%coinF%" from "%addressF%" address to "%addressF)
+                    c srcAddr dstAddr
                 -- TODO: this should be removed in production
                 let txHash = hash tx
+                whenJust mRems $ \(_, remCAddr) -> do
+                    () <$ addHistoryTx remCAddr curr title desc (THEntry txHash tx True Nothing)
+                    removeAccount srcCAddr
                 () <$ addHistoryTx dstCAddr curr title desc (THEntry txHash tx False Nothing)
                 addHistoryTx srcCAddr curr title desc (THEntry txHash tx True Nothing)
 
 getHistory
     :: forall ssc m.
        (SscHelpersClass ssc, WalletWebMode ssc m)
-    => CAddress -> Maybe Word -> Maybe Word -> m ([CTx], Word)
-getHistory cAddr skip limit = do
-    (minit, cachedTxs) <- transCache <$> getHistoryCache cAddr
-
-    TxHistoryAnswer {..} <- flip (untag @ssc getTxHistory) minit
-        =<< decodeCAddressOrFail cAddr
-
-    -- Add allowed portion of result to cache
-    let fullHistory = taHistory <> cachedTxs
-        lenHistory = length taHistory
-        cached = drop (lenHistory - taCachedNum) taHistory
-
-    unless (null cached) $
-        updateHistoryCache cAddr taLastCachedHash taCachedUtxo (cached <> cachedTxs)
-
-    cHistory <- mapM (addHistoryTx cAddr ADA mempty mempty) fullHistory
+    => CWalletAddress -> Maybe Word -> Maybe Word -> m ([CTx], Word)
+getHistory wAddr skip limit = do
+    cAccAddrs <- getWalletAccAddrsOrThrow wAddr
+    cAddrs    <- forM cAccAddrs (decodeCAddressOrFail . caaAddress)
+    cHistory <- do
+        (minit, cachedTxs) <- transCache <$> getHistoryCache wAddr
+
+        TxHistoryAnswer {..} <- untag @ssc getTxHistory cAddrs minit
+
+        -- Add allowed portion of result to cache
+        let fullHistory = taHistory <> cachedTxs
+            lenHistory = length taHistory
+            cached = drop (lenHistory - taCachedNum) taHistory
+
+        unless (null cached) $
+            updateHistoryCache wAddr taLastCachedHash taCachedUtxo
+            (cached <> cachedTxs)
+
+        -- TODO [CSM-931]: what should this line do?
+        mapM (addHistoryTx undefined ADA mempty mempty) fullHistory
```

Comment:
Actually, that's `undefined` I introduced in last commit with changing `getHistory` argument from `CAccountAddress` to `CWalletAddress` :smile: